mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -825,42 +859,116 @@ class WidgetAdapter(
+                    val buttonView = buttonViews.getOrPut(Position(row, column)) {
+                        val newButton = spareViews.removeFirstOrNull()
+                            ?: initData.inflater.inflate(
+                                R.layout.widgetlist_sectionswitchitem_button,
+                                table,
+                                false
+                            ) as MaterialButton
+
+                        table.addView(
+                            newButton,
+                            GridLayout.LayoutParams(
+                                GridLayout.spec(row, GridLayout.FILL, 1f),
+                                GridLayout.spec(column, GridLayout.FILL, 1f)
+                            )
+                        )
+                        newButton
+                    }
+
                     // Rows and columns start with 1 in Sitemap definition, thus decrement them here
-                    val mapping = mappings.firstOrNull { it.row - 1 == row && it.column - 1 == column }
+                    val button = buttons.firstOrNull { (it.row ?: 0) - 1 == row && (it.column ?: 0) - 1 == column }
                     // Create invisible buttons if there's no mapping so each cell has an equal size
-                    buttonView.isInvisible = mapping == null
-                    if (mapping != null) {
+                    buttonView.isInvisible = button == null
+                    if (button != null) {
+                        buttonView.tag = button
                         buttonView.setOnClickListener(this)
-                        buttonView.setTextAndIcon(connection, mapping)
-                        buttonView.tag = mapping.value
+                        buttonView.setOnTouchListener(this)
+                        buttonView.setTextAndIcon(
+                            connection = connection,
+                            label = button.label,
+                            iconRes = button.icon,
+                            labelColor = button.labelColor,
+                            iconColor = button.iconColor,
+                            mapper = colorMapper
+                        )
+
+                        //
+                        // 1. button.stateless == true is NOT the logical inverse of button.stateless == false
+                        //    because stateless == null means stateless
+                        //    so don't use `stateless == true` by rearranging the clauses below
+                        //
+                        // 2. Rule: isChecked cannot be changed when isCheckable is false
+                        //    and it will be stuck at whatever state it was before.
+                        //    So:
+                        //    before setting the checked state, set checkable to true
+                        //    consequently,
+                        //    before setting checkable to false, clear the checked state first
+                        //
```

Comment:
The comment seems a little bit verbose ;-) Rewording suggestion:

```
if (button.stateless == false) {
    // stateful button: make checkable and set checked state afterwards (isChecked can not be set if isCheckable is false)
    buttonView.isCheckable = true
    buttonView.isChecked = button.item?.state?.asString == button.command
} else {
    // stateless button: not checkable (unset isChecked before isCheckable for the reason outlined above)
    buttonView.isChecked = false
    buttonView.isCheckable = false
}
```

Reply:
Thanks! :)